src/postProposalVersions.ts

```diff
@@ -0,0 +1,320 @@
+          opportunityId,
+          externalId: proposalExternalId,
+        },
+        {
+          timeout: requestTimeoutMs,
+          headers,
+        },
+      )
+    ).data;
+  } else {
+    console.log(`Found existing proposal with id=${proposal.id} for opportunityId=${opportunityId}, applicantId=${applicant.id}, and externalId=${proposalExternalId}`);
+  }
+  return proposal;
+};
+
+const postProposalVersion = async (
+  proposal: Proposal,
+  form: ApplicationForm,
+  row: CsvRow,
+) => {
+  const proposalVersion: ProposalVersion = {
+    proposalId: proposal.id,
+    applicationFormId,
+    fieldValues: [],
+  };
+  Object.keys(row).forEach((key) => {
+    // Where the CSV header name matches the label in the application field form, add a value.
+    const formFields: ApplicationFormField[] = form.fields.filter((field) => field.label === key);
+    // We can replicate the same application form field to multiple canonical fields.
+    // The specific use case is "Organization Legal Name" to "Organization Name".
+    Object.values(formFields).forEach((formField) => {
+      const fieldValue = row[key];
+      if (fieldValue === undefined) {
+        throw new Error(`Undefined value for key '${key}' in row`);
+      } else if (fieldValue !== null && fieldValue !== '') {
+        const proposalValue: ProposalFieldValue = {
+          applicationFormFieldId: formField.id,
+          position: formField.position,
+          value: fieldValue,
+        };
+        proposalVersion.fieldValues.push(proposalValue);
+      } else {
+        console.log(`Field value for '${key}' for proposal '${proposal.id}' was null or empty: skipped.`);
+      }
+    });
+    if (formFields.length === 0) {
+      console.log(`Failed to find any field label matching '${key}' in row '${jsonSubstring(row)}...', got ${formFields.length}`);
+    }
+  });
+
+  // Post the new proposal version using the data in `proposalVersion`.
+  await axios.post<ProposalVersion>(
+    `${apiUrl}/proposalVersions`,
+    proposalVersion,
+    {
+      timeout: requestTimeoutMs,
+      headers,
+    },
+  );
+};
+
+const postProposalVersions = async () => {
+  let rowsRead = 0;
+  let versionsPosted = 0;
+
+  try {
+    const form = (
+      await axios.get<ApplicationForm>(`${apiUrl}/applicationForms/${applicationFormId}`, {
+        headers,
+        params: {
+          includeFields: 'true',
+        },
+      })
+    ).data;
+
+    console.log(`Got the relevant application form: '${jsonSubstring(form)}...'`);
+
+    const proposals = (
+      await axios.get<Bundle<Proposal>>(`${apiUrl}/proposals`, {
+        headers,
+        params: {
+          _page: 1,
+          _count: 100000,
+        },
+      })
+    ).data.entries;
+
+    console.log(`Got ${proposals.length} existing proposals.`);
+
+    const csvParser = csvParse({
+      columns: true,
+    });
+    csvInput.pipe(csvParser);
+
+    // Type of `row` is still `any` here:
+    for await (const csvRow of csvParser) {
```

Comment:
Here I am using your suggestion "in particular its async iterator api: https://csv.js.org/parse/api/async_iterator/" which when reading the code is much more elegant. Unfortunately, airbnb disagrees?

Reply:
Yeah, I gave you bad advice - sorry about that! I recently went through something [similar](https://github.com/PhilanthropyDataCommons/data-scripts/pull/24#discussion_r1206113952).

We have time set aside tomorrow to pair on fixing this (and other things), but right now I'd suggest switching over to the [sync version](https://csv.js.org/parse/api/sync/).

Reply:
After fixing a couple of things we set up the sync version. Instead of an async iterator for loop, we used `csvParser.map` wrapped in a `Promise.all` to allow for use of `async/await` throughout. Many thanks, Jason!